// @algorandfoundation/tealscript/dist/lib/contract.js
"use strict";
/* eslint-disable max-classes-per-file */
/* eslint-disable class-methods-use-this */
/* eslint-disable no-unused-vars */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable no-undef */
/// <reference path="../../types/global.d.ts" />
Object.defineProperty(exports, "__esModule", { value: true });
exports.PendingGroup = void 0;
class PendingGroup {
    /**
     * Adds ABI method to the pending transaction group. The two type arguments in combination with the
     * name argument are used to form the the method signature to ensure typesafety.
     *
     * @typeParam ArgsType - A tuple type corresponding to the types of the method arguments
     * @typeParam ReturnType - The return type of the method
     *
     * @param params - The parameters of the method call
     *
     */
    addMethodCall(params) { }
    addPayment(params) { }
    addAppCall(params) { }
    addAssetTransfer(params) { }
    addAssetCreation(params) { }
    addOnlineKeyRegistration(params) { }
    addOfflineKeyRegistration(params) { }
    addAssetConfig(params) { }
    addAssetFreeze(params) { }
    submit() { }
}
exports.PendingGroup = PendingGroup;
class Contract {
    /**
     * The method called when creating the application. The default create method will
     * allow the contract to be created via a bare NoOp appcall and throw an error if called
     * with any arguments.
     */
    createApplication(...args) {
        if (args.length > 0) {
            throw Error();
        }
    }
    /**
     * The method called when attempting to update the application. The default update method will
     * always throw an error
     */
    updateApplication(...args) {
        throw Error();
    }
    /**
     * The method called when attempting to delete the application. The default delete method will
     * always throw an error
     */
    deleteApplication(...args) {
        throw Error();
    }
    /**
     * The method called when an account opts-in to the application. The default opt-in method will
     * always throw an error
     */
    optInToApplication(...args) {
        throw Error();
    }
    /**
     * The method called when an account closes out their local state. The default close-out method
     * will always throw an error
     */
    closeOutOfApplication(...args) {
        throw Error();
    }
    /**
     * The method called when an account clears their local state. The default ClearState
     * method does nothing. ClearState will always allow a user to delete their local state,
     * reagrdless of logic.
     */
    clearState() { }
}
exports.default = Contract;
